adapters/rhythmone/rhythmone.go

```diff
@@ -0,0 +1,148 @@
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("bad server response: %d. ", err),
+		}}
+	}
+
+	var errs []error
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(5)
+
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &sb.Bid[i],
+				BidType: getMediaTypeForImp(sb.Bid[i].ImpID, internalRequest.Imp),
+			})
+		}
+	}
+	return bidResponse, errs
+}
+
+func getMediaTypeForImp(impId string, imps []openrtb.Imp) openrtb_ext.BidType {
+	mediaType := openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impId {
+			if imp.Banner != nil {
+				mediaType = openrtb_ext.BidTypeBanner
+			} else if imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeVideo
+			}
+			return mediaType
+		}
+	}
+	return mediaType
+}
+
+func NewRhythmoneBidder(endpoint string) *RhythmoneAdapter {
+	return &RhythmoneAdapter{
+		endPoint: endpoint,
+	}
+}
+
+func (a *RhythmoneAdapter) preProcess(req *openrtb.BidRequest, errors []error) (*openrtb.BidRequest, string, []error) {
+	numRequests := len(req.Imp)
+
+	requestImpCopy := req.Imp
+	var uri string = ""
+	for i := 0; i < numRequests; i++ {
+		imp := requestImpCopy[i]
+		var bidderExt adapters.ExtImpBidder
+		err := json.Unmarshal(imp.Ext, &bidderExt)
+		if err != nil {
+			err = &errortypes.BadInput{
+				Message: fmt.Sprintf("ext data not provided in imp id=%s. Abort all Request", imp.ID),
+			}
+			errors = append(errors, err)
+			return nil, "", errors
+		}
+		var rhythmoneExt openrtb_ext.ExtImpRhythmone
+		err = json.Unmarshal(bidderExt.Bidder, &rhythmoneExt)
+		if err != nil {
+			err = &errortypes.BadInput{
+				Message: fmt.Sprintf("placementId | zone | path not provided in imp id=%s. Abort all Request", imp.ID),
+			}
+			errors = append(errors, err)
+			return nil, "", errors
+		}
+		rhythmoneExt.S2S = true
+		rhythmoneExtCopy, err := json.Marshal(&rhythmoneExt)
+		if err != nil {
+			errors = append(errors, err)
+			return nil, "", errors
+		}
+		bidderExtCopy := openrtb_ext.ExtBid{
+			Bidder: rhythmoneExtCopy,
+		}
+		impExtCopy, err := json.Marshal(&bidderExtCopy)
+		if err != nil {
+			errors = append(errors, err)
+			return nil, "", errors
+		}
+		imp.Ext = impExtCopy
+		req.Imp[i] = imp
```

Comment:
This could be dangerous, but we do supply each adapter with its own shallow copy of the request imps, so this line will not cause any conflicts. So this is good. :)